config/sampling.go

```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		return nil, err
+	}
+	if localSampledSamplers != nil {
+		pbc.LocalSampled = localSampledSamplers.ActiveSampler
+		for id, config := range localSampledSamplers.Samplers {
+			samplers[id] = config
+		}
+	}
+
+	localNotSampledSampler, err := ConvertSamplerToConfig(p.LocalNotSampled)
+	if err != nil {
+		return nil, err
+	}
+	if localNotSampledSampler != nil {
+		pbc.LocalNotSampled = localNotSampledSampler.ActiveSampler
+		for id, config := range localNotSampledSampler.Samplers {
+			samplers[id] = config
+		}
+	}
+
+	samplers[sampling.ParentBasedID] = sampling.SamplerConfig{
+		SamplerType: sampling.SamplerParentBased,
+		Config:      pbc,
+	}
+
+	return &sampling.Config{
+		Samplers:      samplers,
+		ActiveSampler: sampling.ParentBasedID,
+	}, nil
+}
+
+// DefaultSampler returns a ParentBased sampler with the following defaults:
+//   - Root: AlwaysOn
+//   - RemoteSampled: AlwaysOn
+//   - RemoteNotSampled: AlwaysOff
+//   - LocalSampled: AlwaysOn
+//   - LocalNotSampled: AlwaysOff
+func DefaultSampler() Sampler {
+	return ParentBased{
+		Root:             AlwaysOn{},
+		RemoteSampled:    AlwaysOn{},
+		RemoteNotSampled: AlwaysOff{},
+		LocalSampled:     AlwaysOn{},
+		LocalNotSampled:  AlwaysOff{},
+	}
+}
+
+// NewSamplerFromEnv creates a Sampler based on the environment variables.
+// If the environment variables are not set, it returns a nil Sampler.
+func NewSamplerFromEnv(lookupEnv func(string) (string, bool)) (Sampler, error) {
+	samplerName, ok := lookupEnv(TracesSamplerKey)
+	if !ok {
+		return nil, nil
+	}
+
+	defaultSampler := DefaultSampler().(ParentBased)
+
+	samplerName = strings.ToLower(strings.TrimSpace(samplerName))
+	samplerArg, hasSamplerArg := lookupEnv(TracesSamplerArgKey)
+	samplerArg = strings.TrimSpace(samplerArg)
+
+	switch samplerName {
+	case SamplerNameAlwaysOn:
+		return AlwaysOn{}, nil
+	case SamplerNameAlwaysOff:
+		return AlwaysOff{}, nil
+	case SamplerNameTraceIDRatio:
+		if hasSamplerArg {
+			ratio, err := strconv.ParseFloat(samplerArg, 64)
+			if err != nil {
+				return nil, err
+			}
+			return TraceIDRatio{Fraction: ratio}, nil
+		}
+		return TraceIDRatio{Fraction: 1}, nil
+	case SamplerNameParentBasedAlwaysOn:
+		defaultSampler.Root = AlwaysOn{}
+		return defaultSampler, nil
+	case SamplerNameParsedBasedAlwaysOff:
+		defaultSampler.Root = AlwaysOff{}
+		return defaultSampler, nil
+	case SamplerNameParentBasedTraceIDRatio:
+		if !hasSamplerArg {
+			defaultSampler.Root = TraceIDRatio{Fraction: 1}
+			return defaultSampler, nil
+		}
+		ratio, err := strconv.ParseFloat(samplerArg, 64)
+		if err != nil {
+			return nil, err
+		}
+		defaultSampler.Root = TraceIDRatio{Fraction: ratio}
+		return defaultSampler, nil
+	default:
+		return nil, errors.New("unknown sampler name")
+	}
+}
+
+// ConvertSamplerToConfig converts a Sampler its internal representation.
+func ConvertSamplerToConfig(s Sampler) (*sampling.Config, error) {
```

Comment:
This needs to be unexported. It should not be used by end-users.

```suggestion
// convertSamplerToConfig converts a Sampler its internal representation.
func convertSamplerToConfig(s Sampler) (*sampling.Config, error) {
```

(broken suggestion based on updates needed elsewhere)

Reply:
This change will need the `config` package to be flattened into the `auto` package or the `Sampler` interface will need to be defined and `internal` and aliased here.

I'm in favor of flattening `config` into `auto` (but I haven't completely reviewed that proposal).

Reply:
The reason I created the config package is to avoid cyclic dependencies which happened in the manager (which is aware of the Provider interface from the config package).

Reply:
As discussed during the SIG meeting (and later with @RonFed)
I think we can merge this PR and do this change in a separate PR